rfcs/migrate-to-async-hyper.md

```diff
@@ -0,0 +1,55 @@
+# Migrating docs.rs to asynchronous code and hyper
+
+1. Update `tokio`, `futures` and `rusoto_{s3, core, credential}` to their latest versions (I think Joshua started on this, I'd be happy to help with finishing it)
+2. Make `main` an `async fn` with `#[tokio::main]` to start the tokio runtime, add the `hyper` dependency and migrate the side services to async (By side services I mean the ones in the daemon)
+
+Making the side services async sounds like a larger thing than it actually is, e.g.
+
+```rust
+thread::spawn(move || loop {
+    thread::sleep(Duration::from_secs(60 * 60 * 6));
+    if let Err(e) = github_updater() {
+        error!("Failed to update github fields: {}", e);
+    }
+})
+```
+
+Would become
+
+```rust
+task::spawn(async move {
+    let mut interval = time::interval(Duration::from_secs(60 * 60 * 6));
+    loop {
+        interval.tick().await;
+        if let Err(e) = github_updater().await {
+            error!("Failed to update github fields: {}", e);
+        }
+    }
+})
+```
+
+This actually will do a lot for the server's performance, as the many threads that were doing nothing but sleeping can now be used for useful work. Additionally, for the `release activity updater` this is a good change, as instead un-sleeping every minute and checking if it's `23:55`, we can use tokio's `interval_at` to make it automatically wake once at `23:55` and execute.
+Additionally, all the services are quite small and consist of basically http requests, so migrating can be literally as simple as adding awaits in some cases.
+
+3. Now, at this point the next step would be to migrate from `iron` to `hyper`, and there's a few ways to go about it
+    1. Move all routing to `hyper`, separate the handlers out from being within `iron` code and forwarding requests with `spawn_blocking`
+        1. Requires some glue code, but adding that (in an inert state where it isn't used) can be done in separate prs
+        2. Will be slightly slower until full migration is complete, as we'll have to translations from hyper requests to iron ones and from `IronResult` into the hyper equivalent
+        3. Looking at the code, this will actually be made much easier by the fact that we already have a custom wrapper around managing routes, and we can use that to our advantage
+    2. Use nginx to have two 'separate' servers on different ports that cross-forward things they can't handle
+        1. Messy and kind of impractical, not to mention difficult to develop locally
+        2. Probably the worst option
+    3. Do it all at once
+        1. Has the greatest code churn, but rips off the band-aid, so to speak
```

Comment:
Another downside is, if this migration doesn't turn to be viable much more development and review time will be wasted compared to just migrating parts of the code first.

For example, crates.io has a custom web framework backed by [a C++ web framework](https://github.com/civetweb/civetweb): when they tried to migrate the underlying framework to Hyper they were forced to roll back due to excessive resource usage. This was a while ago and shouldn't be taken as a "don't switch to Hyper", but still, we *might* have to revert this.

Reply:
Yes, that is a definite concern. Hopefully though there won't be any adverse side effects to switching over